### PR TITLE
feat(lora): per-module rank/alpha overrides via rank_pattern / alpha_…

### PR DESCRIPTION
**Pull request by thad0ctor**
# Description

Adds two optional config fields, `lora_rank_pattern` and `lora_alpha_pattern`, that forward to PEFT's LoraConfig.rank_pattern` / `LoraConfig.alpha_pattern`. The memory-efficient LoRA merger now honors them per-module so merging matches training.

Also fixes a **pre-existing dead AdaLoRA-detection check** in the same merger (`target_rank` → `target_r` / `peft_type == "ADALORA"`); PEFT never writes `target_rank`, so the old gate was effectively off.

New `docs/lora.qmd` introduces the feature with an example.

## Motivation and Context

Plain LoRA forces uniform rank/alpha across all targeted modules. Per-module overrides are useful for: asymmetric attention vs MLP ranks, multimodal towers (vision vs language), or fine-tuning specific layers more heavily. PEFT has supported this for years, but axolotl had no YAML path and its memory-efficient merger actively rejected such adapters as "AdaLoRA".

## How has this been tested?

**Unit / integration**: `tests/utils/lora/` — **49 pass** (8 new). Covers schema validation (empty dict, non-int rejection, `PositiveInt` bound), `load_lora(config_only=True)` round-trip into PEFT `LoraConfig`, regex resolver matching, the AdaLoRA-heuristic regression, end-to-end `alpha_pattern` merge, and DoRA + `alpha_pattern`.

**End-to-end vs PEFT, real gradient-trained adapter**: trained a LoRA on `HuggingFaceTB/SmolLM2-135M` with `alpha_pattern={"layers.0.self_attn.q_proj": 64}` (global `alpha=16`, `r=8`) for 5 real `AdamW` steps (loss `6.492 → 5.707`), saved the adapter, then merged via both PEFT's `merge_and_unload()` and `axolotl.merge_lora_sharded_efficient`. Across 4 sampled layers: **max-abs-diff = 0.0** (bit-exact). Merged model reloads via `AutoModelForCausalLM.from_pretrained` and forward-passes cleanly.

**Full CI scope** (pre-commit + pytest): all lint hooks pass (ruff, ruff-format, mypy, bandit). Pytest across `tests/` (excluding e2e), `tests/patched/`, `tests/cli/`, `tests/monkeypatch/`, `tests/integrations/`: **1,407 passed, 0 regressions**.

## AI Usage Disclaimer

Claude (Anthropic, Opus 4.7) user to prepare and test. All changes reviewed by the human author before commit.

## Types of changes

- [x] New feature (non-breaking)
- [x] Bug fix — pre-existing dead AdaLoRA-detection check
- [ ] Breaking change
- [ ] Documentation update

## Summary by CodeRabbit

* **New Features**
  * Added configuration options for per-module LoRA rank and alpha pattern overrides
  * LoRA merge now honors per-module rank and alpha settings without additional flags

* **Documentation**
  * Added comprehensive guide for LoRA per-module configuration with usage patterns and examples

* **Tests**
  * Added extensive test coverage for pattern validation, configuration propagation, and merge behavior

[Review Change Stack](https://app.coderabbit.ai/change-stack/axolotl-ai-cloud/axolotl/pull/3673?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Review by ved1beta — APPROVED** [review]
NICE, lgtm

**Comment by thad0ctor** [comment]
> NICE, lgtm

happy to help!

File: docs/lora.qmd [hunk 1]
```diff
@@ -0,0 +1,52 @@
+---
+title: "LoRA"
```
Comment (NanoCode012) [inline docs/lora.qmd:2]: You should add this file to `.quarto` so that the doc builds this file.

**Review by NanoCode012 — COMMENTED** [review]
Thanks, this looks nice. Have you done some test runs with this?

File: docs/lora.qmd (hunk 1 above)
Comment (thad0ctor) [reply docs/lora.qmd:2]: Addressed via commit I just pushed, I'll respond to your testing question momentarily @NanoCode012

**Comment by thad0ctor** [comment]
> Thanks, this looks nice. Have you done some test runs with this?

@NanoCode012 Yes, locally:

  1. Local test on Qwen3.5-0.8B (RTX 3090) 
  Trained with global r=8, alpha=16 and pattern overrides r=32, alpha=128 on layer-3 attention only. Confirmed via three checks: adapter_config.json persists both dicts; saved safetensors show per-module rank (layer-3 lora_A=(32,1024), layer-7 (8,1024)); PEFT runtime scaling is 4.0 on layer-3 vs 2.0 on layer-7 **proves alpha actually flows through and isn't silently using the global.**
  
```
   base_model: Qwen3.5-0.8B
  adapter: lora
  lora_r: 8
  lora_alpha: 16
  lora_dropout: 0.0
  lora_target_linear: true
  lora_rank_pattern:
    ".*\\.layers\\.3\\..*(q_proj|k_proj|v_proj|o_proj)$": 32
  lora_alpha_pattern:
    ".*\\.layers\\.3\\..*(q_proj|k_proj|v_proj|o_proj)$": 128
  sequence_len: 256
  micro_batch_size: 1
  gradient_accumulation_steps: 1
  max_steps: 5
  learning_rate: 1e-4
  bf16: true
  save_strategy: steps
  save_steps: 5
  save_total_limit: 1
```
  Second validation run with 5090:
  
  - Training complete and model saved successfully
  - Verify script ran on the saved adapter: rank_pattern PASS, alpha_pattern PASS (layer-3 r=32/scaling=4.0, layer-7 r=8/scaling=2.0)


  2. SmolLM2-135M (merger equivalence):
  Trained 5 real AdamW steps with alpha_pattern={"layers.0.self_attn.q_proj": 64} (global
  alpha=16, r=8), loss 6.492 → 5.707. Merged the saved adapter via both PEFT's
  merge_and_unload() and axolotl's merge_lora_sharded_efficient. Max-abs-diff across 4
  sampled layers = 0.0 (bit-exact). **Merged model reloads and forward-passes cleanly.**

```
  base_model: HuggingFaceTB/SmolLM2-135M
  adapter: lora
  lora_r: 8
  lora_alpha: 16
  lora_alpha_pattern:
    "layers.0.self_attn.q_proj": 64
  # (no lora_rank_pattern in this test — pure alpha override)
  max_steps: 5
  optim: adamw_*  

```

**Comment by thad0ctor** [comment]
> Thanks, this looks nice. Have you done some test runs with this?

@NanoCode012 

I also merge tested to Qwen 3.5 0.8B:

| Item | Value |
|---|---|
| Base model | Qwen3.5-0.8B |
| Global LoRA | r=8, alpha=16 |
| `rank_pattern` | `{layer-3 attn → 32}` |
| `alpha_pattern` | `{layer-3 attn → 128}` |
| Merger A (reference) | PEFT `merge_and_unload()` |
| Merger B (under test) | fork's `merge_lora_sharded_efficient` (default `memory_efficient` path) |
| Comparison | tensor-by-tensor max-abs-diff on fp16 weights |

## Results

| Tensor / module | max-abs-diff | Note |
|---|---|---|
| layer-3 q/k/v/o_proj (**patterned**, r=32, α=128) | **0.0** | bit-identical |
| layer-7 q/k/v_proj (unpatterned, r=8, α=16) | **0.0** | bit-identical |
| layer-7 o_proj | 2.4e-7 | 1 bf16 ULP |
| layer-5 mlp.{gate,up,down}_proj | **0.0** | bit-identical |
| **Overall across 320 shared tensors** | **3.8e-6** | 1 bf16 ULP |
| LoRA-targeted weights exact zero | 165/186 | |
| LoRA-targeted weights with non-zero diff | 21/186 | all ≤3.8e-6, on `linear_attn.{in_proj_z, out_proj}` |
| Forward pass on merged model | PASS | clean, non-NaN logits |

The same adapter was merged two ways and the resulting weights diffed tensor-by-tensor. The patterned layer-3 attention modules came out bit-identical. Everything else matched within one bf16 ULP (overall max 3.8e-6), and the merged model forward-passes cleanly.

The 21 sub-ULP non-zero diffs are all concentrated on Qwen3.5's `linear_attn` sub-projections (`in_proj_z`, out_proj`) and trace to a matmul-order difference inside the bf16 noise floor.

The patterned modules came out more exact than the unpatterned ones, opposite of what a pattern bug would produce. Combined with the SmolLM2-135M test (which covered alpha_pattern` only and was also bit-exact), this confirms end-to-end that the memory-efficient merger correctly honors both `rank_pattern` and `alpha_pattern` per-module.

**Review by NanoCode012 — CHANGES_REQUESTED** [review]
Forgot to add, you'll need to add to the quarto for the doc to appear

File: tests/utils/lora/test_merge_lora.py [hunk 2]
```diff
@@ -331,6 +332,125 @@ def test_sharded_efficient_merge(self, tmp_path):
...
+        )
+
+        merge_lora_sharded_efficient(
+            base_model_path=model_dir,
+            lora_adapter_path=adapter_dir,
+            output_path=tmp_path / "output",
+            device="cpu",
+        )
+
+    def test_merge_applies_alpha_pattern_per_module(self, tmp_path):
```
Comment (NanoCode012) [inline tests/utils/lora/test_merge_lora.py:412]: Maybe let's add a test for `rank_pattern`?

File: src/axolotl/loaders/adapter.py [hunk 3]
```diff
@@ -333,6 +337,13 @@ def load_lora(
 
     if cfg.lora_model_dir:
         LOG.debug("Loading pretrained PEFT adapter")
+        if cfg.lora_rank_pattern or cfg.lora_alpha_pattern:
+            # PeftModel.from_pretrained uses the adapter's saved config, so cfg-side patterns would be silently dropped.
+            LOG.warning(
+                "lora_rank_pattern/lora_alpha_pattern are ignored when loading an "
+                "existing adapter via lora_model_dir; the saved adapter_config.json "
```
Comment (NanoCode012) [inline src/axolotl/loaders/adapter.py]: Hmm, what if someone is continue training an adapter, they can load via lora model dir and train further?

**Comment by thad0ctor** [comment]
> Forgot to add, you'll need to add to the quarto for the doc to appear

all set

File: tests/utils/lora/test_merge_lora.py (hunk 2 above)
Comment (thad0ctor) [reply tests/utils/lora/test_merge_lora.py:412]: Added two end-to-end merge tests: `rank_pattern` alone (q_proj at rank 16 vs global 8, each module merged at its own `alpha/r` scale) and `rank_pattern` + `alpha_pattern` combined, since the two interact in the scale factor. Per-module rank is inferred from the saved A/B weight shapes, which these tests now verify explicitly.

[8ec3770](https://github.com/axolotl-ai-cloud/axolotl/pull/3673/commits/8ec37709b2f0cabc04dbcbe4fe3d64d1314599d3)

File: src/axolotl/loaders/adapter.py (hunk 3 above)
Comment (thad0ctor) [reply src/axolotl/loaders/adapter.py]: `PeftModel.from_pretrained` reads the patterns from the adapter's saved `adapter_config.json`, so per-module ranks/alphas persist through continued training; the warning only meant the YAML values can't override the saved ones.

I updated the warning though in 
8ec3770, the warning is now mismatch-aware: it compares cfg patterns against the saved `adapter_config.json` and only fires when they genuinely differ. 

Smoke-tested both paths: `resume_from_checkpoint` rebuilds from YAML and honors patterns (a changed rank fails loudly with a shape mismatch), and `lora_model_dir` with matching YAML loads silently with saved patterns governing. Also documented both behaviors in a short "Resuming and continuing training" section in docs/lora.qmd.

**Review by NanoCode012 — APPROVED** [review]

**Comment by NanoCode012** [comment]
Thank you

**Comment by thad0ctor** [comment]
> Thank you

You're welcome
